More careful handling of GPS NAV data

ljbade opened the pull request:
Fixes #254 and #159.

So far I have implemented:
- IODE and IODC cross checks. Needs testing for when they don't match (e.g. cut over)
- URA extraction and checking for 15 unhealthy value, also works out the value in meters and stores in ephemeris struct. Could be used for weighting in the position solution. Needs testing, so far I have only seen satellites with 2.0 or 2.8m URA.
- all 6 bits of health flags are correctly checked. Currently any unhealthy signal will mark entire satellite as unhealthy in the stored ephemeris. Could be made smarter to only check for unhealthy L1 C/A.
- curve fit interval is correctly calculated and used. Needs testing via simulator for cases when GPS sats go into extended operation mode (e.g. major catastrophe takes out both GPS control centres)

Questions I have:
- we should use IODE instead of IODC for comparing ephemeris (since IODC is unique within a week, IODE only 6 hours) - what code needs to be changed here?
- I have added some new members to the ephemeris structs, does the binary interface need updating?
- how can I modify `nav_msg.c` to extract and store per sat the alert flag, separate from ephemeris struct (since this could change between ephemeris updates)?
- how can I modify the position calculation to ignore sats with an alert or unhealthy bits

@cbeighley Who is best to help me with these?

swiftnav-jenkins commented:
Can one of the admins verify this patch?

ljbade commented:
One of the python tests is failing, and I'm not sure why.

ljbade commented:
I wonder if the alert flag should also be part of the observation messages?

mookerji commented:
Hey @ljbade, 

- The ephemeris test breaking is a unit test of the libswiftnav bindings on the Ephemeris object. If the semantics of the `ephemeris_good` function have changed, then the test should be updated accordingly.
- By "binary interface" do you mean the fields of the libsbp ephemeris message? If so, the answer is largely Yes. The rover and the base station exchange ephemeris messages over the radio, and we also consume them externally off the device. If those fields are only used internally, then don't bother. 
- For the position solution stuff, I recommend reading [calc_PVT](https://github.com/swift-nav/libswiftnav/blob/master/src/pvt.c) and adding an exclusion there. We currently handle exclusions for RAIM, so that might be a good example to crib off of.

ljbade commented:
@mookerji For the alert flag, I did find this bit of code: https://github.com/swift-nav/piksi_firmware/commit/9d0e7bb66a96a2ef822899082608384f2d9d0c5c#diff-3db690123986aa7cf0d7d32c546cebe0R498

I extended the exclusion of unhealthy ephemeris sats to include the alert flag...

As for `libsbp` I have added to the ephemeris.h `iodc` `ura` and `fit_interval`. There was already a field in `libsbp` for IODC (and the python bindings) but it was not filled out...

So I think I only need `fit_interval` at this stage. `ura` would be needed if we use that down the road in the position calculation to give lower priority to less accurate sats.

mookerji commented:
@ljbade , well, if you need to update libsbp, I updated our versioning policy here: https://github.com/swift-nav/libsbp/pull/322. 

The specification for stable messages is fixed. You'll need to add a new SBP message type for an ephemeris with additional fields, and regenerate the C, Python, and Haskell bindings (our most frequently used bindings). Afterwards, propagate them to the firmware. My recommendation is to add all the remaining fields (`fit_interval` and `ura`) to the ephemeris message if you intend on using them in the future. Adding new messages causes a bit of churn in our tools, so best to do that only once.

That being said, the last time we audited the message definitions, we either overlooked the fields you mentioned (from the IDC?) or decided not to include them. I don't remember which one. @fnoble @henryhallam ?

mfine commented:
Right, the ephemeris structures between libswiftnav and libsbp are getting mapped in [sbp_utils.c#L267-L329]](https://github.com/swift-nav/piksi_firmware/blob/master/src/sbp_utils.c#L267-L329). If the SBP ephermeris gets extended, it'll need a new msg and a rev of the library for compatibility. Pretty straightforward - the ephemeris has been extended a number of times already (see  the already deprecated ephemeris messages).

ljbade commented:
@mfine Would it be worth waiting for a review to see what is needed to handle the recently added SBAS ephemeris before creating a new message definition?

I think we can merge this PR and come back to update `libsbp` in a later PR.

ljbade commented:
I have changed the ephemeris comparison functions to use IODC (as well as IODE).

I have also added checks for the alert flag to the relevant firmware branch.

Next thing is to work on the test cases.

ljbade commented:
I have updated the `ephemeris_equal` to handle the new GPS and SBAS fields.

I have also updated `test_ephemeris` to take into account changes around `fit_interval`.

I also want to implement a few more in there I created TODOs for. But my Python is rusty.

Can I update the `eph` object to change its fields directly, or do I have to create a whole new object?

ljbade commented:
The python test is still failing on `assert eph.good(g.GpsTime(**{ 'wn': 1867, 'tow': 518400.0,}))`

@mookerji Are you able to take a closer look at why this is still failing?

mfine commented:
@ljbade that sounds good - bundling up the changes around a message format update makes a lot of sense - it's not a painfree process.

mookerji commented:
@ljbade yeah, I can check it out in a bit.

mookerji commented on python/tests/test_ephemeris.py:
```diff
@@ -44,6 +46,12 @@ def test_sat_state():
                                   'af1': 3.637978807091713e-12,
                                   'af2': 0.0}})
   assert eph.good(g.GpsTime(**{ 'wn': 1867, 'tow': 518400.0,}))
+  eph.healthy = 0
```
Ack! Figured out what was going on here. 

checkout https://github.com/swift-nav/libswiftnav/blob/tidy-gps-nav/python/swiftnav/ephemeris.pyx#L28.

You'll need to add `fit_interval` to be initialized. Currently, it gets set to the initialized value, which is 0, so this test fails.

Lastly, you can't set properties for the Ephemeris object, since the property setter isn't defined. Consequently, this next line at 49 will fail. You could inline another Ephemeris object with `healthy` set to 0, or you can define property setters for those fields.

mookerji replied:
I can't believe I'd say this but maybe we should bite the bullet and migrate to using Swig.

ljbade commented:
retest this please

swiftnav-jenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://zazu.ci.swiftnav.com/job/libswiftnav%20Pull%20Requests/400/
Test PASSed.

ljbade commented:
This should be ready for review.

ljbade commented:
retest this please

swiftnav-jenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://zazu.ci.swiftnav.com/job/libswiftnav%20Pull%20Requests/402/
Test PASSed.

ljbade commented:
@jacobmcnamee or @gsmcmullin Can you review this please?

gsmcmullin commented on src/ephemeris.c:
```diff
@@ -257,7 +254,65 @@ u8 ephemeris_good(const ephemeris_t *eph, gps_time_t t)
+*/
+float decode_ura_index(const u8 index) {
+  static float values[16] = {
+    2.0f,    // 0
```
Consider using c99 style: `[0] = 2.0f,`, or just removing the comments showing the array indices.

ljbade replied:
Ah yeah I will do it like that.

gsmcmullin commented:
This looks great.
Also, thanks for adding some tests for the xyz ephemeris.

ljbade commented:
@gsmcmullin I made the change to the URA table.

Let me know if you now happy to merge.

gsmcmullin commented:
Go ahead and merge.

ljbade commented:
Hmm I just noticed that I overlooked table Table 20-XII for calculation of URA. Making note for me to come and amend that.
